src/net/sourceforge/kolmafia/request/GenericRequest.java

```diff
@@ -1851,18 +1851,21 @@ private boolean handleServerRedirect() {
         return false;
       }
 
-      String oldpwd = GenericRequest.passwordHashValue;
+      String oldpwd = this.getFormField("pwd");
```

Comment:
is this necessarily non-null?

Reply:
Interesting.

I was testing this by timing in with "pull 1 star" - which is a GET request that requires a pwd.
I was - sometimes - getting a storage.php request with no pwd, because the current passwordHash was "".
KoL would respond with a 404.

I also tested with the Refresh Session button, which submits api.php, no pwd. But that is not a GET request.

You are correct - getFormField can return a null if the field is not there. But we only look at oldpwd with GET requests. The original code looked at GenericRequest.passwordHash from before the timein - and if there is none, that is "". Doing a SingleStringReplace of that string with a new password hash doesn't exactly work. I fixed that.

I can fix that, but the question is why I was sometimes getting a storage request with no pwd field?
I can't reproduce that any more.

I need to figure out how that happened, and go back to setting oldpwd to the original password hash - which can never be null.

This is very irritating debugging.

Reply:
This happens:
```
> timeout
> pull 1 line
(KoLmafia requests a storage request, which is a GET with a pwd)
(we time in and replace the old pwd with the current one)
(the pull succeeds)
```
and now this:
```
timeout
> pull 1 line
(KoLmafia requests a storage request, which is a GET with a pwd)
(we time in and reject the ping, remaining logged out - and password hash cleared)
(we do not resubmit the pull)
> pull 1 line
(KoLmafia requests a storage request, which is a GET - but no pwd, since we have none.)
(we time in and accept the ping. We have a new password hash)
(we resubmit the request - still with no password hash, since we didn't know to replace it.)
(KoL returns a 200, but, since there was no pwd, did not actually pull anything.)
```

Basically, if you choose to refuse the ping during a timein - or a relogin - submitting a request that requires a pwd while you are logged out will time in, bu the request will fail.

I'm not sure how to fix that. Somehow remember an old pwd when you timein/relogin and reject a pin, so subsequent newly constructed requests with a pwd will have something (invalid) which we can replace with the new one?

That sounds messy - and I want to work on that in its own PR.

This one fixes your observation: it only looks for a pwd in the URL if GenericRequest.passwordHash is empty - and if the field does not exist in the URL (i.e. getFormField returns null), no potential NPE.

Reply:
OK. I fixed that by giving you a passwordHash of "bogus" if you reject a ping test and we log you out.
This will make the URL submitted while logged out have a bogus pwd, which we will replace.